jobs/base_model_validator.py

```diff
@@ -0,0 +1,203 @@
+
+        Yields:
+            beam.pvalue.TaggedOutput. An element of the output PCollection for
+            the doFn which represents an error as a key value pair.
+        """
+        element = self.clone(model)
+        regex_string = self.regex_string
+
+        if not re.compile(regex_string).match(element.id):
+            yield errors.IdModelValidationError(element)
+
+
+class ValidateDeleted(BaseValidatorDoFn):
+    """DoFn to check whether models marked for deletion are stale."""
+
+    def process(self, model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+            model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            beam.pvalue.TaggedOutput. An element of the output PCollection for
+            the doFn which represents an error as a key value pair.
+        """
+        element = self.clone(model)
+        date_now = datetime.datetime.utcnow()
+
+        date_before_which_models_should_be_deleted = (
+            date_now -
+            cron_services.PERIOD_TO_HARD_DELETE_MODELS_MARKED_AS_DELETED
+        )
+
+        period_to_hard_delete_models_in_days = (
+            cron_services.PERIOD_TO_HARD_DELETE_MODELS_MARKED_AS_DELETED.days)
+
+        if element.last_updated < date_before_which_models_should_be_deleted:
+            yield errors.StaleDeletedModelValidationError(element)
+
+
+class ValidateModelTimeFields(BaseValidatorDoFn):
+    """DoFn to check whether created_on and last_updated timestamps are
+    valid."""
+
+    def process(self, model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+            model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            beam.pvalue.TaggedOutput. An element of the output PCollection for
+            the doFn which represents an error as a key value pair.
+        """
+
+        element = self.clone(model)
+        if element.created_on > (element.last_updated + MAX_CLOCK_SKEW_SECS):
+            yield errors.TimeFieldModelValidationError(element)
+
+        current_datetime = datetime.datetime.utcnow()
+        if element.last_updated > current_datetime:
+            yield errors.CurrentTimeModelValidationError(element)
+
+
+class BaseModelValidator(beam.PTransform):
+    """Composite Beam Transform which returns a pipeline of validation
+    errors."""
+
+    def expand(self, model_pipe):
+        """Function that takes in a beam.PCollection of datastore models and
+        returns a beam.PCollection of validation errors.
+
+        Args:
+            model_pipe: beam.PCollection. A collection of models.
+
+        Returns:
+            beam.PCollection. A collection of errors represented as
+            key-value pairs.
+        """
+        not_deleted, deleted = (model_pipe | beam.Map(
+            self._check_deletion_status)
```

Comment:
Let's inline this function as a lambda, seems trivial enough to deserve it:

```suggestion
            lambda m: beam.pvalue.TaggedOutput(
                'deleted' if m.deleted else 'not_deleted', m))
```